pkg/operator/staticpod/controller/prune/prune_controller.go

```diff
@@ -0,0 +1,223 @@
+	c.queue.AddRateLimited(dsKey)
+
+	return true
+}
+
+func (c *PruneController) sync() error {
+	_, operatorStatus, _, err := c.operatorConfigClient.Get()
+	if err != nil {
+		return err
+	}
+
+	err = c.pruneRevisionHistory(operatorStatus)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getPrunerPodName(podName, nodeName string) string {
+	return fmt.Sprintf("prune-%s-%s", podName, nodeName)
+}
+
+func (c *PruneController) pruneRevisionHistory(operatorStatus *operatorv1.StaticPodOperatorStatus) error {
+	for _, nodeStatus := range operatorStatus.NodeStatuses {
+		// Read status config map to figure which deployments to keep
+		statusConfigMap, err := c.kubeClient.CoreV1().ConfigMaps(c.targetNamespace).Get(getStatusConfigMapName(nodeStatus.TargetRevision, nodeStatus.NodeName), metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		// store IDs as strings to join for protectedIDs param to pruner pod
+		var succeededRevisionIDs []string
+		var failedRevisionIDs []string
+
+		for revisionID, revisionStatus := range statusConfigMap.Data {
+			switch revisionStatus {
+			case string(corev1.PodSucceeded):
+				succeededRevisionIDs = append(succeededRevisionIDs, revisionID)
+			case string(corev1.PodFailed):
+				failedRevisionIDs = append(failedRevisionIDs, revisionID)
+			}
+		}
+		sort.Strings(succeededRevisionIDs)
+		sort.Strings(failedRevisionIDs)
+
+		err = c.pruneEligibleIDs(succeededRevisionIDs, failedRevisionIDs, c.succeededRevisionLimit, "succeeded", nodeStatus.NodeName)
+		if err != nil {
+			return err
+		}
+		err = c.pruneEligibleIDs(failedRevisionIDs, succeededRevisionIDs, c.failedRevisionLimit, "failed", nodeStatus.NodeName)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *PruneController) pruneEligibleIDs(revisionIDs, protectedRevisionIDs []string, revisionLimit int, podName, nodeName string) error {
+	// only prune if enough failed/succeeded revisions to prune
+	if len(revisionIDs) > revisionLimit {
+		// ids should be sorted, so looking back `revisionLimit` ids from the last element should give max id
+		maxEligibleRevision := revisionIDs[len(revisionIDs)-1-revisionLimit]
+		if err := c.ensurePrunePod(podName, nodeName, maxEligibleRevision, protectedRevisionIDs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *PruneController) ensurePrunePod(podName, nodeName, maxEligibleRevision string, protectedRevisions []string) error {
+	required := resourceread.ReadPodV1OrDie([]byte(prunerPod))
+	required.Name = getPrunerPodName(podName, nodeName)
+	required.Namespace = c.targetNamespace
+	required.Spec.NodeName = nodeName
+	required.Spec.Containers[0].Image = c.prunerPodImageFn()
+	required.Spec.Containers[0].Command = c.command
+	required.Spec.Containers[0].Args = append(required.Spec.Containers[0].Args,
+		fmt.Sprintf("-v=%d", 4),
+		fmt.Sprintf("--max-eligible-id=%s", maxEligibleRevision),
+		fmt.Sprintf("--protected-ids=%s", strings.Join(protectedRevisions, ",")),
```

Comment:
> This looks at risk of exceeding the CLI limit.

What limit you're talking about, the OS one is pretty high and I don't see any in cobra library.

Reply:
I agree with @deads2k that this could probably be handled better, with very high history limits we could be passing a lot here (at least with the current approach which runs 2 pods: one to prune successful revisions, where *all* failed revisions are "protected" followed by a pod to prune failed revisions, where *all* successful revisions are "protected")

Reply:
How many protected you're imagining we'll have?

Reply:
The max would be the history limit set by the operator, if the pruner is running successfully. So it could be 10, if the history limit is 10, but if the history limit is 50 then we're passing a lot more. I don't know what, if any, the CLI limit is though